File: implementations/go/aep/governance_test.go
package aep

import "testing"

func TestAuthorizeGrantsStandardActionsToApprovedRoles(t *testing.T) {
	for _, tc := range []struct {
		roles  []string
		action string
	}{
		{[]string{"viewer"}, "event.subscribe"},
		{[]string{"publisher"}, "event.publish"},
		{[]string{"publisher"}, "task.submit"},
		{[]string{"operator"}, "task.manage"},
		{[]string{"tenant-admin"}, "governance.audit.read"},
		{[]string{"cross-tenant-operator"}, "tenant.cross_access"},
	} {
		decision := Authorize(PolicyRequest{ActorID: "agent-1", ActorTenant: "tenant-a", Roles: tc.roles, Action: tc.action, TargetTenant: "tenant-a"})
		if decision != (PolicyDecision{Allowed: true, Action: tc.action, Reason: "allowed", ActorTenant: "tenant-a", TargetTenant: "tenant-a"}) {
			t.Fatalf("unexpected decision for %s: %#v", tc.action, decision)
		}
	}
}

func TestAuthorizeDeniesActionWithoutApprovedRole(t *testing.T) {
	decision := Authorize(PolicyRequest{ActorID: "agent-1", ActorTenant: "tenant-a", Roles: []string{"viewer"}, Action: "event.publish", TargetTenant: "tenant-a"})
	want := PolicyDecision{Allowed: false, Action: "event.publish", Reason: "missing_role", ActorTenant: "tenant-a", TargetTenant: "tenant-a"}
	if decision != want {
		t.Fatalf("got %#v, want %#v", decision, want)
	}
}

func TestAuthorizeDeniesRequestsMissingActorIdentityOrTenantContext(t *testing.T) {
	missingIdentity := Authorize(PolicyRequest{ActorTenant: "tenant-a", Roles: []string{"publisher"}, Action: "event.publish", TargetTenant: "tenant-a"})
	missingTenant := Authorize(PolicyRequest{ActorID: "agent-1", ActorTenant: "tenant-a", Roles: []string{"publisher"}, Action: "event.publish"})
	if missingIdentity.Reason != "missing_identity" {
		t.Fatalf("expected missing identity denial, got %#v", missingIdentity)
	}
	if missingTenant.Reason != "missing_tenant" {
		t.Fatalf("expected missing tenant denial, got %#v", missingTenant)
	}
}

func TestAuthorizeDeniesCrossTenantActionsWithoutCrossTenantAccess(t *testing.T) {
	denied := Authorize(PolicyRequest{ActorID: "agent-1", ActorTenant: "tenant-a", Roles: []string{"operator"}, Action: "event.publish", TargetTenant: "tenant-b"})
	allowed := Authorize(PolicyRequest{ActorID: "agent-1", ActorTenant: "tenant-a", Roles: []string{"cross-tenant-operator"}, Action: "event.publish", TargetTenant: "tenant-b"})
	if denied.Allowed || denied.Reason != "tenant_access_denied" {
		t.Fatalf("expected cross-tenant denial, got %#v", denied)
	}
	want := PolicyDecision{Allowed: true, Action: "event.publish", Reason: "allowed", ActorTenant: "tenant-a", TargetTenant: "tenant-b"}
	if allowed != want {
		t.Fatalf("got %#v, want %#v", allowed, want)
	}
}

File: implementations/typescript/packages/governance/test/governance.test.js
import assert from "node:assert/strict";
import test from "node:test";
import { authorize } from "../src/index.js";

const actor = (roles, tenantId = "tenant-a") => ({
  actorId: "agent-1",
  tenantId,
  roles
});

test("grants the standard actions to their approved roles", () => {
  for (const [roles, action] of [
    [["viewer"], "event.subscribe"],
    [["publisher"], "event.publish"],
    [["publisher"], "task.submit"],
    [["operator"], "task.manage"],
    [["tenant-admin"], "governance.audit.read"],
    [["cross-tenant-operator"], "tenant.cross_access"]
  ]) {
    const decision = authorize({ actor: actor(roles), action, targetTenant: "tenant-a" });
    assert.deepEqual(decision, {
      allowed: true,
      action,
      reason: "allowed",
      actorTenant: "tenant-a",
      targetTenant: "tenant-a"
    });
  }
});

test("denies an action without an approved role", () => {
  const decision = authorize({
    actor: actor(["viewer"]),
    action: "event.publish",
    targetTenant: "tenant-a"
  });

  assert.deepEqual(decision, {
    allowed: false,
    action: "event.publish",
    reason: "missing_role",
    actorTenant: "tenant-a",
    targetTenant: "tenant-a"
  });
});

test("denies requests missing actor identity or tenant context", () => {
  const missingIdentity = authorize({
    actor: { tenantId: "tenant-a", roles: ["publisher"] },
    action: "event.publish",
    targetTenant: "tenant-a"
  });
  const missingTenant = authorize({
    actor: actor(["publisher"]),
    action: "event.publish",
    targetTenant: null
  });

  assert.equal(missingIdentity.reason, "missing_identity");
  assert.equal(missingTenant.reason, "missing_tenant");
});

test("denies cross-tenant actions unless the actor has cross-tenant access", () => {
  const denied = authorize({
    actor: actor(["operator"]),
    action: "event.publish",
    targetTenant: "tenant-b"
  });
  const allowed = authorize({
    actor: actor(["cross-tenant-operator"]),
    action: "event.publish",
    targetTenant: "tenant-b"
  });

  assert.equal(denied.allowed, false);
  assert.equal(denied.reason, "tenant_access_denied");
  assert.deepEqual(allowed, {
    allowed: true,
    action: "event.publish",
    reason: "allowed",
    actorTenant: "tenant-a",
    targetTenant: "tenant-b"
  });
});
